runtime/v2/runhcs/process.go

```diff
@@ -0,0 +1,167 @@
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package runhcs
+
+import (
+	"context"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/log"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+func newProcess(ctx context.Context, id string, pid uint32, stdin, stdout, stderr string, terminal bool) (*process, error) {
+	ioPipes, err := newPipeSet(ctx, stdin, stdout, stderr, terminal)
+	if err != nil {
+		return nil, err
+	}
+	process := &process{
+		id:       id,
+		pid:      pid,
+		stdin:    stdin,
+		stdout:   stdout,
+		stderr:   stderr,
+		terminal: terminal,
+		io:       ioPipes,
+	}
+	return process, nil
+}
+
+type process struct {
+	sync.Mutex
+
+	id  string
+	pid uint32
+
+	stdin    string
+	stdout   string
+	stderr   string
+	terminal bool
+	io       *pipeSet
+
+	// started track if the process has ever been started and will not be reset
+	// for the lifetime of the process object.
+	started bool
+
+	exitWg sync.WaitGroup
+	// If nil exitCode and exitTime are not available. Values will only be set
+	// after a return from exitWg.Wait() or the caller MUST hold p.Lock() to
+	// check or update.
+	exitCode *uint32
+	exitTime *time.Time
+}
+
+// closeIO closes the stdin of the executing process to unblock any waiters
+func (p *process) closeIO() {
+	p.Lock()
+	defer p.Unlock()
+
+	p.io.stdin.Close()
+}
+
+// safeExitCode will return safeExitCodeLocked() after taking the lock.
+func (p *process) safeExitCode() (uint32, time.Time) {
+	p.Lock()
+	defer p.Unlock()
+	return p.safeExitCodeLocked()
+}
+
+// safeExitCodeLocked will return the process exit code, exit time if the
+// process has exited, otherwise it will return 255, time 0. This call will not
+// block waiting for the process to exit.
+//
+// The caller MUST hold p.Lock() OR be called after p.exitWg.Wait().
+func (p *process) safeExitCodeLocked() (uint32, time.Time) {
+	if p.exitCode != nil {
+		return *p.exitCode, *p.exitTime
+	}
+	return 255, time.Time{}
+}
+
+func (p *process) start(ctx context.Context) error {
+	p.Lock()
+	defer p.Unlock()
+
+	if p.started {
+		return errors.Wrap(errdefs.ErrFailedPrecondition, "process already started")
+	}
+
+	cmd := exec.Command(
+		runhcsBinary,
+		runhcsDebug,
+		"start",
+		p.id)
+
+	cmd.Stdin = p.io.stdin
+	cmd.Stdout = p.io.stdout
+	cmd.Stderr = p.io.stderr
+
+	ec, err := startCmd(ctx, cmd)
+	if err != nil {
+		p.io.Close()
+		return err
+	}
+
+	p.exitWg.Add(1)
```

Comment:
I think we may want to use a channel for this.  There would be multiple people waiting on a single container and a channel handles this well.  I cannot remember, but I tried this in the past with a waitgroup and had issues with multiple wait calls

Reply:
This logic is written such that `N` number of waiters can wait by calling `p.waitForExit()`. Once the process's background goroutine actually completes it calls `p.exitWg.Done()` so a process has exactly `1` call to `p.exitWg.Add(1)` done internally and all external waiters are blocked until this is completed. For a call like `Stat(...)` we expose a call `p.safeExitCode()` that is a non-blocking variant of `p.waitForExit()` with the expectation that the `exitCode` and `exitTime` might be invalid if it has not exited `(255, time.Time{})`. I could certainly expose a channel here but I don't see any benefit to the pattern over using wait groups. Am I missing something?

Reply:
ok, no worries

Reply:
Changed based on your feedback.